internal/core/data/coredata_test.go

```diff
@@ -0,0 +1,362 @@
+
+	"github.com/edgexfoundry/edgex-go/internal/core/data/messaging"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/db"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/db/memory"
+	"github.com/edgexfoundry/edgex-go/pkg/clients/logging"
+	"github.com/edgexfoundry/edgex-go/pkg/clients/metadata/mocks"
+	"github.com/edgexfoundry/edgex-go/pkg/models"
+	"github.com/gorilla/mux"
+	"github.com/stretchr/testify/mock"
+	"gopkg.in/mgo.v2/bson"
+)
+
+var testEvent models.Event
+var testRoutes *mux.Router
+
+const (
+	testDeviceName string = "Test Device"
+	testOrigin     int64  = 123456789
+)
+
+// Mock implementation of the event publisher for testing purposes
+type mockEventPublisher struct{}
+
+func newMockEventPublisher(config messaging.PubSubConfiguration) messaging.EventPublisher {
+	return &mockEventPublisher{}
+}
+
+func (zep *mockEventPublisher) SendEventMessage(e models.Event) error {
+	return nil
+}
+
+func TestMain(m *testing.M) {
+	testRoutes = LoadRestRoutes()
+	LoggingClient = logger.NewMockClient()
+	mdc = newMockDeviceClient()
+	ep = newMockEventPublisher(messaging.PubSubConfiguration{})
+	chEvents = make(chan interface{}, 10)
+	os.Exit(m.Run())
+}
+
+// Supporting methods
+// Reset() re-initializes dependencies for each test
+func reset() {
+	Configuration = &ConfigurationStruct{}
+	testEvent.Device = testDeviceName
+	testEvent.Origin = testOrigin
+	testEvent.Readings = buildReadings()
+	dbClient = &memory.MemDB{}
+	testEvent.ID, _ = dbClient.AddEvent(&testEvent)
+}
+
+func newMockDeviceClient() *mocks.DeviceClient {
+	client := &mocks.DeviceClient{}
+
+	mockAddressable := models.Addressable{
+		Address:  "localhost",
+		Name:     "Test Addressable",
+		Port:     3000,
+		Protocol: "http"}
+
+	mockDeviceResultFn := func(id string) models.Device {
+		if bson.IsObjectIdHex(id) {
+			return models.Device{Id: bson.ObjectIdHex(id), Name: testEvent.Device, Addressable: mockAddressable}
+		}
+		return models.Device{}
+	}
+	client.On("Device", mock.MatchedBy(func(id string) bool {
+		return id == "valid"
+	})).Return(mockDeviceResultFn, nil)
+	client.On("Device", mock.MatchedBy(func(id string) bool {
+		return id == "404"
+	})).Return(mockDeviceResultFn, types.ErrNotFound{})
+	client.On("Device", mock.Anything).Return(mockDeviceResultFn, fmt.Errorf("some error"))
+
+	mockDeviceForNameResultFn := func(name string) models.Device {
+		device := models.Device{Id: bson.NewObjectId(), Name: name, Addressable: mockAddressable}
+
+		return device
+	}
+	client.On("DeviceForName", mock.MatchedBy(func(name string) bool {
+		return name == testEvent.Device
+	})).Return(mockDeviceForNameResultFn, nil)
+	client.On("DeviceForName", mock.MatchedBy(func(name string) bool {
+		return name == "404"
+	})).Return(mockDeviceForNameResultFn, types.ErrNotFound{})
+	client.On("DeviceForName", mock.Anything).Return(mockDeviceForNameResultFn, fmt.Errorf("some error"))
+
+	return client
+}
+
+func newMockDb() interfaces.DBClient {
+	DB := &dbMock.DBClient{}
```

Comment:
What if all of this mock db stuff was in a internal/core/data/interfaces/mocks/DBClient_test.go?

Reply:
I think putting anything in a file we expect to be 100% generated is a bad idea.

Reply:
To your point about coredata_test, it's not testing the DBClient, it's mocking it. The reason I went with coredata_test is because the name is literally true, even if it doesn't follow the conventions of the language. This is the opposite: it's literally untrue, but follows the language convention. I'd rather be clear than do this.